q/verbose/verbose.q

```diff
@@ -0,0 +1,471 @@
+    .finos.verbose.xkey[xk] (0!tbl) iasc ?[tbl;();0b;s]
+    };
+
+//wrapper to allow null-ignoring xdesc in secure gateway (raw xasc may modify table in-place)
+.finos.verbose.ndesc:{[sortCols;tbl]
+    if[not type[sortCols] in -11 11h; '"sort columns must be symbol(list)"];
+    if[not .Q.qt[tbl]; '".finos.verbose.ndesc expects a table"];
+    sortCols xdesc tbl};
+
+// unkeys a table. It is just a projection of .finos.verbose.xkey setting the first parameter to empty list
+.finos.verbose.unkey: .finos.verbose.xkey[()];
+
+.finos.verbose.priv.validateSelectArgs:{[tbl;constr;grp;stat;cnt;ord]
+    if[not .Q.qt[tbl]; '".finos.verbose.select expects a table"];
+    if[not 0h=type constr; '"constraints must be a general list"];
+    if[not type[grp] in -11 -1 0h;
+            if[not 99h=type grp; '"invalid type for groupby"];
+            if[not 11h=type key grp; '"groupby must have symbol keys"];
+    ];
+    if[not type[stat] in -11 0h;
+        if[not 99h=type stat; '"invalid type for stat"];
+        if[not 11h=type key stat; '"stat must have symbol keys"];
+    ];
+    if[not -7h=type cnt; '"cnt must be long"];
+    if[not 0h=type ord; '"ord must be a general list"];
+    if[not 2=count ord; '"ord must have size 2"];
+    if[not ord[0] in (<:;>:); '"first element of ord must be <: or >:"];
+    if[not -11h=type ord[1]; '"second element of ord must be a symbol"];
+    };
+
+//wrapper to allow select in secure gateway (the ? operator has some overloads with side-effects)
+.finos.verbose.select:{[tbl;constr;grp;stat]
+    .finos.verbose.priv.validateSelectArgs[tbl;constr;grp;stat;0W;(<:;`i)];
+    ?[tbl;constr;grp;stat]};
+
+.finos.verbose.select5:{[tbl;constr;grp;stat;cnt]
+    .finos.verbose.priv.validateSelectArgs[tbl;constr;grp;stat;cnt;(<:;`i)];
+    ?[tbl;constr;grp;stat;cnt]};
+
+.finos.verbose.select6:{[tbl;constr;grp;stat;cnt;ord]
+    .finos.verbose.priv.validateSelectArgs[tbl;constr;grp;stat;cnt;ord];
+    ?[tbl;constr;grp;stat;cnt;ord]};
+
+.finos.verbose.selectPerfTest:{[tbl;constr;grp;stat]
+    if[not 99h=type stat; '"stat must be a dictionary"];
+    {[tbl;constr;grp;stcol;stexpr].Q.ts[.finos.verbose.select;(tbl;constr;grp;enlist[stcol]!enlist stexpr)][0]}[tbl;constr;grp]'[key stat;value stat]};
+
+//wrapper to allow update in secure gateway (the ! operator has some overloads with side-effects and raw update may modify tables in-place)
+.finos.verbose.update:{[tbl;constr;grp;stat]
+    if[not .Q.qt[tbl]; '".finos.verbose.update expects a table"];
+    if[not 0h=type constr; '"constraints must be a general list"];
+    if[not -1h=type grp;
+        if[not 99h=type grp; '"groupby must be boolean or dictionary"];
+        if[not 11h=type key grp; '"groupby must have symbol keys"];
+    ];
+    if[not()~stat;
+        if[not 99h=type stat; '"stat must be empty list or dictionary"];
+        if[not 11h=type key stat; '"stat must have symbol keys"];
+    ];
+    ![tbl;constr;grp;stat]};
+
+.finos.verbose.cond:{[cond;valTrue;valFalse]
```

Comment:
Missing comments for this one and a few others below.

Reply:
Comments added.